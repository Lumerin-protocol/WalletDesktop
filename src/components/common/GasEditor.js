import withGasEditorState from '@lumerin/wallet-ui-logic/src/hocs/withGasEditorState';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import React from 'react';

import { FieldBtn } from './Btn';
import TextInput from './TextInput';
import Flex from './Flex';
import Sp from './Spacing';

const GasLabel = styled.span`
  opacity: 0.5;
  font-size: 1.3rem;
  white-space: nowrap;
  margin-right: 1em;
`;

const ErrorMsg = styled.div`
  margin-top: 1em;
  color: ${p => p.theme.colors.danger};
`;

class GasEditor extends React.Component {
  static propTypes = {
    gasEstimateError: PropTypes.bool,
    onInputChange: PropTypes.func.isRequired,
    useCustomGas: PropTypes.bool.isRequired,
    onGasToggle: PropTypes.func.isRequired,
    priceError: PropTypes.bool,
    gasPrice: PropTypes.string.isRequired,
    gasLimit: PropTypes.string.isRequired,
    errors: PropTypes.shape({
      gasPrice: PropTypes.string,
      gasLimit: PropTypes.string
    }).isRequired
  };

  render() {
    return (
      <React.Fragment>
        {this.props.useCustomGas ? (
          <Flex.Row justify="space-between">
            <Flex.Item grow="1" basis="0">
              <TextInput
                data-testid="gas-limit-field"
                onChange={this.props.onInputChange}
                label="Gas Limit (Units)"
                error={this.props.errors.gasLimit}
                value={this.props.gasLimit}
                id="gasLimit"
              />
            </Flex.Item>
            <Sp mt={6} mx={2.5} />
            <Flex.Item grow="1" basis="0">
              <TextInput
                data-testid="gas-price-field"
                onChange={this.props.onInputChange}
                label="Gas Price (Gwei)"
                error={this.props.errors.gasPrice}
                value={this.props.gasPrice}
                id="gasPrice"
              />
            </Flex.Item>
          </Flex.Row>
        ) : (
          <Flex.Row justify="space-between" align="flex-start">
            <Flex.Row rowwrap shrink="1">
              <GasLabel>Gas Limit: {this.props.gasLimit} (Units)</GasLabel>
              <GasLabel>Gas Price: {this.props.gasPrice} (Gwei)</GasLabel>
            </Flex.Row>
            <FieldBtn
              data-testid="edit-gas-btn"
              tabIndex="-1"
              onClick={this.props.onGasToggle}
            >
              EDIT GAS
            </FieldBtn>
          </Flex.Row>
        )}
        {this.props.priceError && (
          <ErrorMsg>
            Current gas price could not be retrieved. Using last known price.
          </ErrorMsg>
        )}
        {this.props.gasEstimateError && (
          <ErrorMsg>Gas limit could not be estimated. Using default.</ErrorMsg>
        )}
      </React.Fragment>
    );
  }
}

export default withGasEditorState(GasEditor);
