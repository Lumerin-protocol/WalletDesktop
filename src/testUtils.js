/* eslint-disable require-path-exists/exists */
/* eslint-disable import/no-unresolved */
import { Provider as ClientProvider } from 'lumerin-wallet-ui-logic/src/hocs/clientContext';
import { Provider, createStore } from 'lumerin-wallet-ui-logic/src/store';
import { render, Simulate } from 'react-testing-library';
import { ThemeProvider } from 'styled-components';
import { MemoryRouter } from 'react-router';
import createClient from './client';
import { merge } from 'lodash';
import theme from 'lumerin-wallet-ui-logic/src/theme';
import React from 'react';

import config from '../config';

/*
 * The same render method of 'react-testing-library' but wrapped with
 * Redux store Provider and 'styled-components' ThemeProvider.
 *
 * element      : the React element you want to wrap
 * initialState : optional state to initialize the Redux store
 *
 * Returns the same object as render() but with an additional 'store'
 * property useful for dispatching actions inside your tests.
 */
export function reduxRender(element, initialState) {
  const client = createClient(config, createStore, initialState);

  const renderResult = render(
    <ClientProvider value={client}>
      <Provider store={client.store}>
        <ThemeProvider theme={theme}>
          <React.Fragment>{element}</React.Fragment>
        </ThemeProvider>
      </Provider>
    </ClientProvider>
  );
  return { ...renderResult, store: client.store };
}

/*
 * The same as reduxRender() above but also wrapped with 'react-router'
 * in order to test navigation flows.
 *
 * Returns the same object as reduxRender()
 */
export function routerRender(element, initialState, initialRoute = '/') {
  return reduxRender(
    <MemoryRouter initialEntries={[initialRoute]}>{element}</MemoryRouter>,
    initialState
  );
}

/*
 * Declaratively test complex form validations
 *
 * getter   : function that receives a data-testid and returns an HTML element
 * formId   : data-testid of the form element
 * formData : object of type { [data-testid]: any } to populate the form
 * errors   : object of type { [data-testid]: String } with expected errors
 */
export function testValidation(getter, formId, { formData, errors }) {
  // Populate the specified form fields
  Object.keys(formData).forEach((fieldId, i) => {
    const field = getter(fieldId);
    field.value = formData[fieldId];
    Simulate.change(field);
  });

  // Submit the form
  Simulate.submit(getter(formId));

  // Check expected errors
  Object.keys(errors).forEach((fieldId, i) => {
    expect(getter(`${fieldId}-error`).textContent).toEqual(
      expect.stringContaining(errors[fieldId])
    );
  });
}

/*
 * Add dataset property (data-foo attribute values) to nodes
 *
 * This hack is required because jsdom has no support for dataset yet
 * which prevents us from testing components that rely on it.
 * See https://github.com/jsdom/jsdom/issues/961
 *
 * element   : DOM node to add the dataset to
 * dataAttrs : attribute names to add to the node
 *
 * Usage:
 * <button data-testid="btn" data-modal="buy" data-foo="bar">Click</button>
 *
 * const btn = withDataset(getByTestId('btn'), 'modal', 'foo')
 * // btn.dataset === { modal: 'buy', foo: 'bar'}
 *
 */
export function withDataset(element, ...dataAttrs) {
  element.dataset = dataAttrs.reduce((acc, attr) => {
    acc[attr] = element.getAttribute(`data-${attr}`);
    return acc;
  }, {});
  return element;
}

/**
 * Return UTC timestamps relative to the current time
 * Useful for setting up different auction scenarios.
 */
export const twoWeeksAgo = () => Date.now() / 1000 - 60 * 60 * 24 * 7 * 2;
export const oneHourAgo = () => Date.now() / 1000 - 60 * 60;
export const inOneHour = () => Date.now() / 1000 + 60 * 60;
export const inOneWeek = () => Date.now() / 1000 + 60 * 60 * 24 * 7;

/*
 * Returns a common initial state for Redux store that is useful for most tests
 * Accepts an optional object with overrides that will be deeply merged with
 * the base state object.
 */
export function getInitialState(overrides = {}) {
  const baseState = {
    config,
    connectivity: { isOnline: true },
    blockchain: { height: 1, gasPrice: config.DEFAULT_GAS_PRICE },
    converter: {
      status: {
        availableEth: '100',
        availableLmr: '100',
        currentPrice: '10'
      }
    },
    auction: {
      status: {
        nextAuctionStartTime: inOneHour(),
        tokenRemaining: '1',
        currentAuction: 10,
        currentPrice: '33000000000',
        genesisTime: twoWeeksAgo()
      }
    },
    session: { isLoggedIn: true },
    rates: { ETH: { token: 'ETH', price: 1 } },
    wallets: {
      syncStatus: 'up-to-date',
      active: 'foo',
      allIds: ['foo'],
      byId: {
        foo: {
          addresses: {
            '0x15dd2028C976beaA6668E286b496A518F457b5Cf': {
              token: {
                [config.LMR_TOKEN_ADDR]: { balance: '5000000000000000000000' }
              },
              balance: '5000000000000000000000',
              transactions: []
            }
          }
        }
      }
    }
  };
  return merge({}, baseState, overrides);
}

/*
 * A sample transaction object. Useful for populating state in tests.
 */
export function getDummyTransaction() {
  return {
    meta: {
      lumerin: {},
      tokens: {
        [config.LMR_TOKEN_ADDR]: {
          event: 'Transfer',
          processing: false,
          value:
            '0x000000000000000000000000000000000000000000000001e5b8fa8fe2ac0000'
        }
      }
    },
    receipt: {
      blockHash: '0x1234567890abcdef',
      blockNumber: 407574,
      cumulativeGasUsed: 38484,
      gasUsed: 38484,
      transactionHash: '0x1234567890abcdef',
      transactionIndex: 0
    },
    transaction: {
      transactionIndex: 0,
      blockNumber: 407574,
      blockHash: '0x1234567890abcdef',
      hash: '0x1234567890abcdef',
      value: '0',
      gasPrice: '5000000000',
      gas: 76968,
      from: '0x15dd2028C976beaA6668E286b496A518F457b5Cf',
      to: '0xde806D6efD432CDeE42573760682D99eDEdC1d89'
    }
  };
}
