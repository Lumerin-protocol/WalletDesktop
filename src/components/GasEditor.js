import { TextInput, FieldBtn, Flex, Sp } from './common'
import { sendToMainProcess, weiToGwei } from '../utils'
import * as selectors from '../selectors'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import config from '../config'
import React from 'react'

const GasLabel = styled.span`
  opacity: 0.5;
  font-size: 1.3rem;
  white-space: nowrap;
  margin-right: 1em;
`

class GasEditor extends React.Component {
  static propTypes = {
    networkGasPrice: PropTypes.string.isRequired,
    useCustomGas: PropTypes.bool.isRequired,
    gasPrice: PropTypes.string.isRequired,
    gasLimit: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    dispatch: PropTypes.func.isRequired,
    errors: PropTypes.shape({
      gasPrice: PropTypes.string,
      gasLimit: PropTypes.string
    }).isRequired
  }

  static initialState = currency => ({
    useCustomGas: false,
    gasPrice: weiToGwei(config.DEFAULT_GAS_PRICE),
    gasLimit: config[`${currency}_DEFAULT_GAS_LIMIT`]
  })

  componentDidMount() {
    // Avoid getting current price if using custom price
    if (this.props.useCustomGas) return

    // Start using the the cached gas price in store
    this.props.onChange({
      target: { id: 'gasPrice', value: weiToGwei(this.props.networkGasPrice) }
    })

    // But also fetch current gas price in background
    sendToMainProcess('get-gas-price', {})
      .then(({ gasPrice }) => {
        gasPrice = parseFloat(gasPrice) < config.DEFAULT_GAS_PRICE
            ? config.DEFAULT_GAS_PRICE
            : gasPrice

        this.props.dispatch({ type: 'gas-price-updated', payload: gasPrice })
        this.props.onChange({
          target: { id: 'gasPrice', value: weiToGwei(gasPrice) }
        })
      })
      .catch(err => console.warn('Gas price request failed', err)) // eslint-disable-line no-console
  }

  onGasToggle = () => {
    const { useCustomGas, onChange } = this.props
    onChange({ target: { id: 'useCustomGas', value: !useCustomGas } })
  }

  render() {
    const { useCustomGas, gasPrice, gasLimit, errors, onChange } = this.props

    return useCustomGas ? (
      <Flex.Row justify="space-between">
        <Flex.Item grow="1" basis="0">
          <TextInput
            data-testid="gas-limit-field"
            onChange={onChange}
            label="Gas Limit (Units)"
            error={errors.gasLimit}
            value={gasLimit}
            id="gasLimit"
          />
        </Flex.Item>
        <Sp mt={6} mx={2.5} />
        <Flex.Item grow="1" basis="0">
          <TextInput
            data-testid="gas-price-field"
            onChange={onChange}
            label="Gas Price (Gwei)"
            error={errors.gasPrice}
            value={gasPrice}
            id="gasPrice"
          />
        </Flex.Item>
      </Flex.Row>
    ) : (
      <Flex.Row justify="space-between" align="flex-start">
        <Flex.Row rowwrap shrink="1">
          <GasLabel>Gas Limit: {gasLimit} (Units)</GasLabel>
          <GasLabel>Gas Price: {gasPrice} (Gwei)</GasLabel>
        </Flex.Row>
        <FieldBtn
          data-testid="edit-gas-btn"
          tabIndex="-1"
          onClick={this.onGasToggle}
        >
          EDIT GAS
        </FieldBtn>
      </Flex.Row>
    )
  }
}

const mapStateToProps = state => ({
  networkGasPrice: selectors.getNetworkGasPrice(state)
})

export default connect(mapStateToProps)(GasEditor)
