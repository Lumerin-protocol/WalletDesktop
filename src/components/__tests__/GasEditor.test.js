import * as testUtils from '../../testUtils'
import { Simulate } from 'react-testing-library'
import GasEditor from '../common/GasEditor'
import React from 'react'

const element = (
  <GasEditor
    onInputChange={jest.fn()}
    useCustomGas
    gasLimit="21000"
    gasPrice="1"
    errors={{}}
  />
)

describe('<GasEditor/>', () => {
  it('should match its snapshot', () => {
    const { container } = testUtils.reduxRender(element)
    expect(container).toMatchSnapshot()
  })
})

export function runValidateTests(el, initialState, formTestId) {
  it('displays an error if GAS LIMIT is not provided', async () => {
    const { getByTestId } = testUtils.reduxRender(el, initialState)
    Simulate.click(getByTestId('edit-gas-btn'))
    testUtils.testValidation(getByTestId, formTestId, {
      formData: { 'gas-limit-field': '' },
      errors: { 'gas-limit-field': 'Gas limit is required' }
    })
  })

  it.skip('displays an error if GAS LIMIT is an invalid value', async () => {
    const { getByTestId } = testUtils.reduxRender(element, initialState)
    Simulate.click(getByTestId('edit-gas-btn'))
    testUtils.testValidation(getByTestId, formTestId, {
      formData: { 'gas-limit-field': 'foo' },
      errors: { 'gas-limit-field': 'Invalid value' }
    })
  })

  it.skip('displays an error if GAS LIMIT is not an integer', async () => {
    const { getByTestId } = testUtils.reduxRender(element, initialState)
    Simulate.click(getByTestId('edit-gas-btn'))
    testUtils.testValidation(getByTestId, formTestId, {
      formData: { 'gas-limit-field': '1.1' },
      errors: { 'gas-limit-field': 'Gas limit must be an integer' }
    })
  })

  it.skip('displays an error if GAS LIMIT is negative', async () => {
    const { getByTestId } = testUtils.reduxRender(element, initialState)
    Simulate.click(getByTestId('edit-gas-btn'))
    testUtils.testValidation(getByTestId, formTestId, {
      formData: { 'gas-limit-field': '-1' },
      errors: { 'gas-limit-field': 'Gas limit must be greater than 0' }
    })
  })

  it.skip('displays an error if GAS PRICE is not provided', async () => {
    const { getByTestId } = testUtils.reduxRender(element, initialState)
    Simulate.click(getByTestId('edit-gas-btn'))
    testUtils.testValidation(getByTestId, formTestId, {
      formData: { 'gas-price-field': '' },
      errors: { 'gas-price-field': 'Gas price is required' }
    })
  })

  it.skip('displays an error if GAS PRICE is an invalid value', async () => {
    const { getByTestId } = testUtils.reduxRender(element, initialState)
    Simulate.click(getByTestId('edit-gas-btn'))
    testUtils.testValidation(getByTestId, formTestId, {
      formData: { 'gas-price-field': 'foo' },
      errors: { 'gas-price-field': 'Invalid value' }
    })
  })

  it.skip('displays an error if GAS PRICE is lower than 1', async () => {
    const { getByTestId } = testUtils.reduxRender(element, initialState)
    Simulate.click(getByTestId('edit-gas-btn'))
    testUtils.testValidation(getByTestId, formTestId, {
      formData: { 'gas-price-field': '0' },
      errors: { 'gas-price-field': 'Gas price can not be lower than 1' }
    })
  })
}
