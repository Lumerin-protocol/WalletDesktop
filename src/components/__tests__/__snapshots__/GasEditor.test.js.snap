// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GasEditor/> should match its snapshot 1`] = `
<div>
  <div
    class="sc-gzVnrw dLNtIY"
  >
    <div
      class="sc-bZQynM jmPdFc"
    >
      <div>
        <label
          class="sc-bxivhb htFESG"
          for="gasLimit"
        >
          Gas Limit (Units)
        </label>
        <input
          class="sc-ifAKCX kSbDeA"
          data-testid="gas-limit-field"
          id="gasLimit"
          type="text"
          value="21000"
        />
      </div>
    </div>
    <div
      class="sc-dnqmqq cnSjSH"
    />
    <div
      class="sc-bZQynM jmPdFc"
    >
      <div>
        <label
          class="sc-bxivhb htFESG"
          for="gasPrice"
        >
          Gas Price (Gwei)
        </label>
        <input
          class="sc-ifAKCX kSbDeA"
          data-testid="gas-price-field"
          id="gasPrice"
          type="text"
          value="1"
        />
      </div>
    </div>
  </div>
</div>
`;
